spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java

```diff
@@ -265,6 +266,7 @@ public void cleanup(@PathVariable("id") Set<Long> ids,
 	 */
 	@RequestMapping(method = RequestMethod.DELETE)
 	@ResponseStatus(HttpStatus.OK)
+	@Async
```

Comment:
This needs to be an opt-in feature and here are a couple of options I can think of:

### Option 1
Add another API (eg. `cleanupAllAsync`). 
- ❌ UI will not pick this up and you must use this via REST API
- ➕ low risk of breaking other users
- ➕ UI does not give the false appearance that all has been cleaned up instantly
- ➕ can add CompletableFuture return value for better usage later from UI/callers etc..

### Option 2
Guard the auto-config w/ `enabled` property - hence w/o `@EnableAsync` the `@Async` is invisible.
- ❌ UI will give false appearance that all has been cleaned up instantly when async is enabled
- ❌ moderate risk of breaking/confusing other users
- ➕ for those users that opt-in, the UI does this async

I am in favor of the latter option. If you agree, please prototype/verify the absence of `@EnableAsync` does what I think it does.

Reply:
I don’t know why it should be an opt in feature as the UI says: „X task will be deleted“ - it doesn’t matter if the user can interact directly after pressing ok (in case of async) or has to wait till the backend processed the operation.

Reply:
I hear you. That is a good point. My concern is 

1. w/o some sort of messaging in the UI that mentions that it will happen in the background I fear some users may start looking around and still see the executions and then try again, then run into an issue, then file bug report, etc.. 

2. whether opt-in or opt-out, we need some sort of "light switch" for the feature in case we run into unforeseen issues in production. 

Because it is late in the release cycle I am just being extra cautious. I do want to get this feature in, but I also am treading carefully.

Reply:
Good point!

As you prefer we can delay this feature as with „clean task executions after n days“ you can run multiple cleanups and shrink down the count of present executions in multiple steps, so there is no need to hurry up with this anymore.😃 

I am going to refactor and add the simplifications and we can think about a good way to implement this in a following release.

WDYT?

Reply:
That would be great and I would appreciate that. I know you put your hard work and effort into this so I wanted to match you w/ whatever effort I could to get this in. Moving it to post 2.11 is the smart decision. Thank you.

Reply:
Just one little question regarding `Option 2`. The current refactoring includes the `ConditionalOnProperty` to check if the functionality is enabled so you have to opt in to use this feature.

From my understanding `@EnableAsync` has to be used with `@Async` as the annotation on the method picks up the thread pool executor provided with `@EnableAsync` - but I am not 100% sure.

Here are some information regarding async implementation: https://spring.io/guides/gs/async-method/ and https://stackoverflow.com/a/53357076

If this is the case I would suggest to still let `@EnableAsync` be on the Configuration, because it is not used when the opt flag in is not set.